build-support/githooks/prepare-commit-msg

```diff
@@ -1,15 +1,66 @@
 #!/usr/bin/env bash
-
 COMMIT_MSG_FILEPATH=$1
 COMMIT_MSG_SRC=$2
 
-NUM_NON_MD_FILES=$(git status -s --porcelain | grep -cv ".\md$")
+# NB: prepare-commit-msg runs in the context of GIT_WORK_TREE, ie: pwd == REPO_ROOT
+source build-support/common.sh
+MERGE_BASE="$(git_merge_base)"
+
+# NB: We consider all changed files since the merge base, rather than only inspecting the latest
+# commit. With a feature branch, this usually means all the local commits since the last push to
+# the PR.
+#
+# That means that we do not calculate skips based off what is in the overall PR, but rather what is
```

Comment:
This does not track.
1. If my merge base is `upstream/master` and I make my 1st commit with rust changes this hook will not mark this commit as skip rust which is good.
2. I then make an additional local commit which has no rust changes. My merge base has not changed, so both this commit and the prior are checked for rust edits. Since the prior commit has edits, this commit is not marked for rust skip.
3. etc, ... all future commits are "poisoned" by the 1st and we get no rust skips in any of the pushes to this PR.

Reply:
I think only a comment change is needed, its fine that we have this poison case. We still get the benefit of skipping rust when the PR commit series never modifies rust. That's a common case.

Reply:
Where I'm confused here is what `merge base` really means? 

In my git workflow, I have `origin` point to `Eric-Arellano/pants` and `upstream` point to `pantsbuild/pants`. For each PR, I checkout a new branch via `git checkout -b ci-skip-rust-tests`. Then, I `git push --set-upstream origin ci-skip-rust-tests`. So, locally, right now, my merge base is `origin/ci-skip-rust-tests` and `git diff --name-only origin/ci-skip-rust-tests` is empty because all my changes have been pushed to `Eric-Arellano/pants` already.

When is it the case that your merge base would be configured to `upstream/master`? Indeed, my comment assumes the merge base will be something like `origin/my-feature-branch`.

Reply:
Ah, I've never heard of that workflow. I label origin and upstream just like you, but set the upstream just once at branch creation time: `git checkout -b mybranch upstream/master`. I then do my local work and `git push origin HEAD`, rinse and repeat until the PR is done.

Reply:
I think mutating upstream is uncommon, but either way, as noted, the poisoned branch is a corner case. You've still got the common case covered.

Reply:
If I underestand this correctly, then I think the poisoned branch behavior is what we want anyway. E.g., commit 1 changes rust, we rebuild and the rust tests fail, and then commit 2 doesn't change rust, we still want to run the rust tests after commit 2, otherwise CI will pass when it should continue to fail.